overlord/snapstate/autorefresh.go

```diff
@@ -489,21 +493,46 @@ func getTime(st *state.State, timeKey string) (time.Time, error) {
 	return t1, nil
 }
 
+func populateRefreshHints(refreshInfo *userclient.PendingSnapRefreshInfo, snapInfo *snap.Info, err *BusySnapError) {
+	for _, appName := range err.AppNames() {
+		if app, ok := snapInfo.Apps[appName]; ok {
+			path := app.DesktopFile()
+			if _, err := os.Stat(path); err == nil {
+				refreshInfo.BusyAppName = appName
+				refreshInfo.BusyAppDesktopEntry = strings.SplitN(filepath.Base(path), ".", 2)[0]
+			}
+		}
+	}
+}
+
 // inhibitRefresh returns an error if refresh is inhibited by running apps.
 //
 // Internally the snap state is updated to remember when the inhibition first
 // took place. Apps can inhibit refreshes for up to "maxInhibition", beyond
 // that period the refresh will go ahead despite application activity.
 func inhibitRefresh(st *state.State, snapst *SnapState, info *snap.Info, checker func(*snap.Info) error) error {
 	if err := checker(info); err != nil {
+		refreshInfo := &userclient.PendingSnapRefreshInfo{
+			InstanceName: info.InstanceName(),
+		}
+		if err, ok := err.(*BusySnapError); ok {
+			populateRefreshHints(refreshInfo, info, err)
+		}
+
 		days := int(maxInhibition.Truncate(time.Hour).Hours() / 24)
 		now := time.Now()
+		client := userclient.New()
 		if snapst.RefreshInhibitedTime == nil {
 			// Store the instant when the snap was first inhibited.
 			// This is reset to nil on successful refresh.
 			snapst.RefreshInhibitedTime = &now
 			Set(st, info.InstanceName(), snapst)
 			if _, ok := err.(*BusySnapError); ok {
+				refreshInfo.TimeRemaining = (maxInhibition - now.Sub(*snapst.RefreshInhibitedTime)).Truncate(time.Second)
+				if err := client.PendingRefreshNotification(context.TODO(), refreshInfo); err != nil {
```

Comment:
I think we should ensure that poking the client is not done synchronously while the state lock being held. Perhaps we need one more wrapped that executes the notification inside a goroutine?

Reply:
I agree, just thinking how to do that correctly. Do you mean as in fire-and-forget?
```
go func() { ... }()
```

Reply:
yes, but maybe in a wrapper, eg. `asyncNotifyPendingRefresh()`, which does the `go ...` inside.

Reply:
Ack, will do

Reply:
Done